python/GafferUI/VectorDataWidget.py

```diff
@@ -1338,6 +1381,113 @@ def _createEditorInternal( self, index ) :
 _Delegate.registerType( IECore.QuatfVectorData.staticTypeId(), _NumericDelegate )
 _Delegate.registerType( IECore.QuatdVectorData.staticTypeId(), _NumericDelegate )
 
+# A delegate that creates a `_NumericDelegate` for each of the color components
+# and an editable color swatch for the color column.
+class _ColorDelegate( _Delegate ) :
+
+	def __init__( self ) :
+
+		_Delegate.__init__( self )
+
+		self.__colorChooser = None
+		self.__popup = None
+
+	def paint( self, painter, option, index ) :
+
+		isColor, value = self.__colorData( index )
+
+		if isColor :
+
+			# in PyQt, option is passed to us correctly as a QStyleOptionViewItemV4,
+			# but in PySide it is merely a QStyleOptionViewItem and we must "cast" it.
+			if hasattr( QtGui, "QStyleOptionViewItemV4" ) :
+				option = QtGui.QStyleOptionViewItemV4( option )
+
+			transformedColor = GafferUI.DisplayTransform.get()( value )
+
+			backgroundBrush = QtGui.QBrush(
+				QtGui.QColor(
+					transformedColor.r * 255.0,
+					transformedColor.g * 255.0,
+					transformedColor.b * 255.0,
+					255.0 if isinstance( value, imath.Color3f ) else value.a * 255.0
+				)
+			)
+
+			painter.fillRect( option.rect, backgroundBrush )
+		else :
+			_Delegate.paint( self, painter, option, index )
+
+	def createEditor( self, parent, option, index ) :
+
+		w = _Delegate.createEditor( self, parent, option, index )
+
+		isColor, value = self.__colorData( index )
+		if isColor :
+			w.setParent( None )
+			# Restore popoup flag after `setParent()` above
+			w.setWindowFlags( QtCore.Qt.Popup )
+
+		return w
+
+	def _createEditorInternal( self, index ) :
+
+		isColor, value = self.__colorData( index )
+
+		if isColor :
+			self.__colorChooser = GafferUI.ColorChooser( value )
+			self.__popup = GafferUI.PopupWindow( "", child = self.__colorChooser )
+
+			# We hand over control of the widget to Qt in `createEditor()`, which
+			# will make the widget visible, but first we have to notify the Gaffer
+			# widget that it's visible to keep them in sync.
+			self.__popup.setVisible( True )
+
+			return self.__popup
+		else :
+			return GafferUI.NumericWidget( value )
+
+	def updateEditorGeometry( self, editor, option, index ) :
+
+		isColor, value = self.__colorData( index )
+		if isColor :
+			center = GafferUI.Widget.mousePosition()
+			size = editor.sizeHint()
+			topLeft = center - imath.V2i( size.width() / 2, size.height() / 2 )
+
+			editor.setGeometry( topLeft.x, topLeft.y, size.width(), size.height() )
```

Comment:
I think this bit might be unnecessary if we called `self.__popup.popup()` in `_createEditorInternal()`?

Reply:
I tried that originally but the popup insists on going to the origin when I don't override `updateEditorGeometry()`. I think that's because it's calling `updateEditorGeometry()` after I was calling `self.__popup.popup()` and getting the original coordinates back again.

I'm not certain, but I think that's happening here : https://github.com/qt/qtbase/blob/ffdd372c7bbda62e9d937f406319f38e3e982774/src/widgets/itemviews/qabstractitemview.cpp#L2731-L2736

Reply:
Here's a version where we don't need to duplicate the `popup()` logic : https://github.com/johnhaddon/gaffer/commit/19e5f818db5d6d6787722838c063145ab7752e0f. The other nice thing about this way is that slots connected to `editSignal()` will be able to create and position windows too...

Reply:
Yes, much nicer - fixed in 882cb927542cff3cab31dbe6e110e937d022efe3